src/main/scala/uk/gov/nationalarchives/omega/api/services/Dispatcher.scala

```diff
@@ -0,0 +1,107 @@
+class Dispatcher(val localProducer: LocalProducer) {
+
+  def run(dispatcherId: Int)(q: Queue[IO, LocalMessage]): IO[Unit] =
+    for {
+      requestMessage <- q.take
+      _ <- IO.delay {
+             println(s"Dispatcher # $dispatcherId, processing message id: ${requestMessage.persistentMessageId}")
+           }
+      serviceRequest                  <- createServiceRequest(requestMessage)
+      businessService                 <- IO.pure(serviceRequest._1)
+      businessServiceRequest          <- IO.pure(serviceRequest._2)
+      validatedBusinessServiceRequest <- validateBusinessServiceRequest(businessService, businessServiceRequest)
+      businessResult                  <- execBusinessService(businessService, validatedBusinessServiceRequest)
+      res                             <- sendResultToJmsQueue(businessResult, requestMessage)
+    } yield res
+
+  private def createServiceRequest(localMessage: LocalMessage): IO[(BusinessService, BusinessServiceRequest)] =
+    localMessage.serviceId match {
+      case ECHO001 =>
+        IO.delay {
+          Tuple2(
+            new EchoService(),
+            EchoRequest(localMessage.messageText)
+          )
+        }
+      // add more service IDs here
+    }
+
+  private def validateBusinessServiceRequest(
+    businessService: BusinessService,
+    businessServiceRequest: BusinessServiceRequest
+  ): IO[ValidatedNec[RequestValidationError, BusinessServiceRequest]] =
+    IO.delay {
+      businessService match {
+        case value: RequestValidation =>
+          value.validateRequest(businessServiceRequest)
+        case _ =>
+          Validated.valid(businessServiceRequest)
+      }
+    }
+
+  private def execBusinessService[T <: BusinessServiceRequest, U <: BusinessServiceResponse, E <: BusinessServiceError](
+    businessService: BusinessService,
+    validatedBusinessServiceRequest: ValidatedNec[RequestValidationError, T]
+  ): IO[ValidatedNec[RequestValidationError, Either[BusinessServiceError, BusinessServiceResponse]]] =
+    IO.delay {
+      validatedBusinessServiceRequest.map(businessService.process)
+    }
+
+  private def sendResultToJmsQueue[U <: BusinessServiceResponse, E <: BusinessServiceError](
+    businessResult: ValidatedNec[RequestValidationError, Either[E, U]],
+    requestMessage: LocalMessage
+  ): IO[Unit] = {
+    val replyMessage: IO[String] = IO.delay {
+      businessResult match {
+        case Valid(Right(businessResult)) =>
+          requestMessage.serviceId match {
+            case ECHO001 => businessResult.asInstanceOf[EchoResponse].text
+            // add more services here
+          }
+
+        case Valid(Left(serviceError)) =>
+          val customerErrorReference = UUID.randomUUID()
+          s"""{status: "SERVICE-ERROR", reference: "$customerErrorReference", code: "${serviceError.code}", message: "${serviceError.message}"}"""
+
+        case Invalid(requestValidationFailures) =>
+          val customerErrorReference = UUID.randomUUID()
```

Comment:
Here we're generating random UUIDs using Java libraries, elsewhere we're using Cat's Generators.

Reply:
This now uses a generator.